### PR TITLE
Refactor komodo insert proposals

### LINKED ISSUE #468: Komodo insert proposals raises AttributeError when updating from version to matrix
The recursive_update() function throws AttributeError with these inputs:
input_release: {"package_a": "version_a"},
upgrade_proposals: {"package_a": {"py38": {"rhel8": "version_b"}}},

### PR DESCRIPTION
**Pull request by jonathan-eq**
Resolves #468

### CODE REVIEW

File: komodo/insert_proposals.py [hunk 1]
```diff
@@ -18,22 +20,41 @@
...
 
 
-def recursive_update(left, right):
-    if right is None:
+def recursive_update(
+    to_be_upgraded: Mapping[str, Union[str, Mapping[str, str]]],
+    upgrades: Optional[Mapping[str, Union[str, Mapping[str, str]]]],
+) -> None:
+    """Updates release in place with upgrades/additions of packages from upgrades.
+        Takes format:
```
Comment (berland) [inline komodo/insert_proposals.py:28]: misaligned whitespace before `Takes`. Could that create issues if this is viewed through sphinx?

File: komodo/insert_proposals.py [hunk 2]
```diff
@@ -18,22 +20,41 @@
...
-            if not d_val:
-                left[key] = value
-            else:
-                recursive_update(d_val, value)
+    for package_name, new_version_or_matrix in upgrades.items():
+        old_version_or_matrix = to_be_upgraded.get(package_name)
+        if isinstance(new_version_or_matrix, Mapping) and isinstance(
+            old_version_or_matrix, Mapping
+        ):
+            # upgrade matrix recursively
```
Comment (berland) [inline komodo/insert_proposals.py:50]: that comment is the same as the line below, can be skipped.

File: komodo/insert_proposals.py (hunk 1 above)
Comment (jonathan-eq) [reply komodo/insert_proposals.py:28]: Formatting docstrings is a nightmare. It wraps if I have it on the same indentation.

File: tests/test_insert_proposals.py [hunk 3]
```diff
@@ -546,3 +546,74 @@ def test_duplicate_package_entry_handling(
...
+                "package_a": {
+                    "py38": {"rhel7": "version_a", "rhel8": "version_b"},
+                    "py312": "version_b",
+                }
+            },
+            id="update_only_one_matrix_version_two_levels_deep",
+        ),
+    ],
+)
+def test_recursive_update(release_dict, upgrade_dict, expected_dict):
```
Comment (berland) [inline tests/test_insert_proposals.py:617]: very nice test!

**Comment by jonathan-eq** [comment]
Quick question. In most places in insert_proposals, we use update and upgrade as the same thing, but it is not necessarily the same thing. I think we should only use update. This includes renaming upgrade_proposals to update_proposals. 
Example: A broken version of a package is added to the release, and we would have to downgrade. This should not be added in a file called `upgrade_proposals.yml`.

File: komodo/insert_proposals.py [hunk 4]
```diff
@@ -192,7 +207,92 @@ def insert_proposals(repo, base, target, git_ref, jobname, joburl) -> None:
...
+    )
+
+    # create new release file
+    from_sha = repo.get_branch(git_ref).commit.sha
+    tmp_ref = repo.create_git_ref(ref="refs/heads/" + tmp_target, sha=from_sha)
+    new_release_file = create_new_release_file(
+        repo, target, new_release_contents, tmp_target
+    )
+
+    # clean the proposal file
```
Comment (berland) [inline komodo/insert_proposals.py:260]: skip comment.

File: komodo/insert_proposals.py [hunk 5]
```diff
@@ -192,7 +207,92 @@ def insert_proposals(repo, base, target, git_ref, jobname, joburl) -> None:
...
+    )
+    return new_release_file
+
+
+def insert_proposals(
+    repo: Repository, base: str, target: str, git_ref: str, jobname: str, joburl: str
+) -> None:
+    tmp_target = target + ".tmp"
+
+    # check that the branches do not already exist
```
Comment (berland) [inline komodo/insert_proposals.py:231]: your functions names are too good to need any comment :)

**Comment by berland** [comment]
Are the lines you have changed covered by the existing tests with a mocked github service?

**Comment by jonathan-eq** [comment]
No, but now that it is easier to test, I will add some good unit tests.

**Comment by jonathan-eq** [comment]
I have added tests now.

**Review by berland — APPROVED** [review]
LGTM

**Pull request merged**